File: R/unigd.R


#' Asynchronous HTTP server graphics device.
#'
#' This function initializes a unigd graphics device and
#' starts a local webserver, that allows for access via HTTP and WebSockets.
#' A link will be printed by which the web client can be accessed using
#' a browser.
#'
#' All font settings and descriptions are adopted from the excellent
#' 'svglite' package.
#'
#' @param width Graphics device width (pixels).
#' @param height Graphics device height (pixels).
#' @param bg Background color.
#' @param pointsize Graphics device point size.
#' @param system_fonts Named list of font names to be aliased with
#'   fonts installed on your system. If unspecified, the R default
#'   families `sans`, `serif`, `mono` and `symbol`
#'   are aliased to the family returned by
#'   [systemfonts::font_info()].
#' @param user_fonts Named list of fonts to be aliased with font files
#'   provided by the user rather than fonts properly installed on the
#'   system. The aliases can be fonts from the fontquiver package,
#'   strings containing a path to a font file, or a list containing
#'   `name` and `file` elements with `name` indicating
#'   the font alias in the SVG output and `file` the path to a
#'   font file.
#' @param reset_par If set to `TRUE`, global graphics parameters will be saved
#'   on device start and reset every time [ugd_clear()] is called (see
#'   [graphics::par()]).
#'
#' @return No return value, called to initialize graphics device.
#'
#' @importFrom systemfonts match_font
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd() # Initialize graphics device and start server
#' ugd_browse() # Or copy the displayed link in the browser
#'
#' # Plot something
#' x <- seq(0, 3 * pi, by = 0.1)
#' plot(x, sin(x), type = "l")
#'
#' dev.off() # alternatively: ugd_close()
#' }
ugd <-
  function(width = getOption("unigd.width", 720),
           height = getOption("unigd.height", 576),
           bg = getOption("unigd.bg", "white"),
           pointsize = getOption("unigd.pointsize", 12),
           system_fonts = getOption("unigd.system_fonts", list()),
           user_fonts = getOption("unigd.user_fonts", list()),
           reset_par = getOption("unigd.reset_par", FALSE)) {

    aliases <- validate_aliases(system_fonts, user_fonts)

    invisible(unigd_ugd_(
      bg, width, height,
      pointsize, aliases,
      reset_par
    ))
  }

stop_if_not_unigd_device <- function(which) {
  if (names(which) != "unigd") {
    stop("Device is not of type unigd. (Start a device by calling: `ugd()`)")
  }
}

#' unigd device status.
#'
#' Access status information of a unigd graphics device.
#' This function will only work after starting a device with [ugd()].
#'
#' @param which Which device (ID).
#'
#' @return List of status variables with the following named items:
#'   `$hsize`: Plot history size (how many plots are accessible),
#'   `$upid`: Update ID (changes when the device has received new information),
#'   `$active`: Is the device the currently activated device.
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' ugd_state()
#' plot(1, 1)
#' ugd_state()
#'
#' dev.off()
#' }
ugd_state <- function(which = dev.cur()) {
  stop_if_not_unigd_device(which)
  return(unigd_state_(which))
}

#' unigd device information.
#'
#' Access general information of a unigd graphics device.
#' This function will only work after starting a device with [ugd()].
#'
#' @param which Which device (ID).
#'
#' @return List of status variables with the following named items:
#'   `$id`: Server unique ID,
#'   `$version`: unigd and library versions.
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' ugd_info()
#'
#' dev.off()
#' }
ugd_info <- function(which = dev.cur()) {
  stop_if_not_unigd_device(which)
  return(unigd_info_(which))
}

#' unigd device renderers.
#'
#' Get a list of available renderers.
#' This function will only work after starting a device with [ugd()].
#'
#' @return List of renderers with the following named items:
#'   `$id`: Renderer ID,
#'   `$mime`: File mime type,
#'   `$ext`: File extension,
#'   `$name`: Human readable name,
#'   `$type`: Renderer type (currently either `plot` or `other`),
#'   `$bin`: Is the file a binary blob or text.
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#'
#' ugd_renderers()
#'
ugd_renderers <- function() {
  unigd_renderers_()
}

#' Query unigd plot IDs
#'
#' Query unigd graphics device static plot IDs.
#' Available plot IDs starting from `index` will be returned.
#' `limit` specifies the number of plots.
#' This function will only work after starting a device with [ugd()].
#'
#' @param index Plot index. If this is set to `0`, the last page will be
#'   selected.
#' @param limit Limit the number of returned IDs. If this is set to a
#'  value > 1 the returned type is a list if IDs. Set to `0` for all.
#' @param which Which device (ID).
#' @param state Include the current device state in the returned result
#'  (see also: [ugd_state()]).
#'
#' @return TODO
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' plot.new()
#' text(.5, .5, "#1")
#' plot.new()
#' text(.5, .5, "#2")
#' plot.new()
#' text(.5, .5, "#3")
#' third <- ugd_id()
#' second <- ugd_id(2)
#' all <- ugd_id(1, limit = Inf)
#' ugd_remove(1)
#' ugd_plot(second)
#'
#' dev.off()
#' }
ugd_id <- function(index = 0, limit = 1, which = dev.cur(), state = FALSE) {
  stop_if_not_unigd_device(which)
  if (is.infinite(limit)) {
    limit <- 0
  }
  res <- unigd_id_(which, index - 1, limit)
  if (state) {
    return(res)
  }
  if (limit == 1 && length(res$plots) > 0) {
    return(res$plots[[1]])
  }
  return(res$plots)
}

page_id_to_index <- function(page, which) {
  if (inherits(page, "unigd_pid")) {
    print(page)
    print(which)
    page <- unigd_plot_find_(which, page$id) + 1
  }
  page
}

#' @export
print.unigd_pid <- function(x, ...) cat(x$id)

#' Render unigd plot and return it.
#'
#' See [ugd_save()] for saving rendered plots as files.
#' This function will only work after starting a device with [ugd()].
#'
#' @param page Plot page to render. If this is set to `0`, the last page will
#'   be selected. Can be set to a numeric plot index or plot ID
#'   (see [ugd_id()]).
#' @param width Width of the plot. If this is set to `-1`, the last width will
#'   be selected.
#' @param height Height of the plot. If this is set to `-1`, the last height
#'   will be selected.
#' @param zoom Zoom level. (For example: `2` corresponds to 200%, `0.5` would
#'   be 50%.)
#' @param as Renderer.
#' @param which Which device (ID).
#'
#' @return Rendered plot. Text renderers return strings, binary renderers
#'   return byte arrays.
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' plot(1, 1)
#' ugd_render(width = 600, height = 400, as = "svg")
#'
#' dev.off()
#' }
ugd_render <- function(page = 0,
                     width = -1,
                     height = -1,
                     zoom = 1,
                     as = "svg",
                     which = dev.cur()) {
  stop_if_not_unigd_device(which)
  page <- page_id_to_index(page, which)
  unigd_render_(which, page - 1, width, height, zoom, as)
}

#' Render unigd plot to a file.
#'
#' See [ugd_render()] for acessing plot data directly in memory without
#' saving as a file.
#' This function will only work after starting a device with [ugd()].
#'
#' @param file Filepath to save plot.
#' @param page Plot page to render. If this is set to `0`, the last page will
#'   be selected. Can be set to a numeric plot index or plot ID
#'   (see [ugd_id()]).
#' @param width Width of the plot. If this is set to `-1`, the last width will
#'   be selected.
#' @param height Height of the plot. If this is set to `-1`, the last height
#'   will be selected.
#' @param zoom Zoom level. (For example: `2` corresponds to 200%, `0.5` would
#'   be 50%.)
#' @param as Renderer. When set to `"auto"` renderer is inferred from the file
#'   extension.
#' @param which Which device (ID).
#'
#' @return Rendered SVG string.
#'
#' @importFrom grDevices dev.cur
#' @importFrom tools file_ext
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' plot(1, 1)
#' ugd_save(file = tempfile(), width = 600, height = 400, as = "png")
#'
#' dev.off()
#' }
ugd_save <- function(file,
                     page = 0,
                     width = -1,
                     height = -1,
                     zoom = 1,
                     as = "auto",
                     which = dev.cur()) {
  stop_if_not_unigd_device(which)
  page <- page_id_to_index(page, which)
  if (as == "auto") {
    as <- tolower(tools::file_ext(file))
    if (!(as %in% ugd_renderers()$id)) {
      stop("Renderer could not automatically be inferred from file extension.",
        " (Set the renderer explicitly with e.g. `ugd_save(..., as = \"svg\")`)")
    }
  }
  ret <- unigd_render_(which, page - 1, width, height, zoom, as)
  if (is.character(ret)) {
    writeLines(text = ret, con = file, useBytes = TRUE)
  } else {
    writeBin(object = ret, con = file)
  }
}

#' Remove a unigd plot page.
#'
#' This function will only work after starting a device with [ugd()].
#'
#' @param page Plot page to remove. If this is set to `0`, the last page will
#'   be selected. Can be set to a numeric plot index or plot ID
#'   (see [ugd_id()]).
#' @param which Which device (ID).
#'
#' @return Whether the page existed (and thereby was successfully removed).
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' plot(1, 1) # page 1
#' hist(rnorm(100)) # page 2
#' ugd_remove(page = 1) # remove page 1
#'
#' dev.off()
#' }
ugd_remove <- function(page = 0, which = dev.cur()) {
  stop_if_not_unigd_device(which)
  if (inherits(page, "unigd_pid")) {
    return(unigd_remove_id_(which, page$id))
  }
  return(unigd_remove_(which, page - 1))
}

#' Clear all unigd plot pages.
#'
#' This function will only work after starting a device with [ugd()].
#'
#' @param which Which device (ID).
#'
#' @return Whether there were any pages to remove.
#'
#' @importFrom grDevices dev.cur
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' plot(1, 1)
#' hist(rnorm(100))
#' ugd_clear()
#' hist(rnorm(100))
#'
#' dev.off()
#' }
ugd_clear <- function(which = dev.cur()) {
  stop_if_not_unigd_device(which)
  return(unigd_clear_(which))
}

#' Close unigd device.
#'
#' This achieves the same effect as [grDevices::dev.off()],
#' but will only close the device if it has the unigd type.
#'
#' @param which Which device (ID).
#' @param all Should all running unigd devices be closed.
#'
#' @return Number and name of the new active device (after the specified device
#'   has been shut down).
#'
#' @importFrom grDevices dev.cur dev.list dev.off
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd()
#' ugd_browse() # open browser
#' hist(rnorm(100))
#' ugd_close() # Equvalent to dev.off()
#'
#' ugd()
#' ugd()
#' ugd()
#' ugd_close(all = TRUE)
#' }
ugd_close <- function(which = dev.cur(), all = FALSE) {
  if (all) {
    ds <- dev.list()
    invisible(lapply(ds[names(ds) == "unigd"], dev.off))
  } else {
    if (which != 1 && names(which(dev.list() == which)) == "unigd") {
      dev.off(which)
    }
  }
}

#' Inline plot rendering.
#'
#' Convenience function for quick inline plot rendering.
#' This is similar to [ugd_render()] but the plotting code
#' is specified inline and an unigd graphics device is managed
#' (created and closed) automatically. Starting a device with [ugd()] is
#' therefore not necessary.
#'
#' @param code Plotting code. See examples for more information.
#' @param page Plot page to render. If this is set to `0`, the last page will
#'   be selected. Can be set to a numeric plot index or plot ID
#'   (see [ugd_id()]).
#' @param width Width of the plot.
#' @param height Height of the plot.
#' @param zoom Zoom level. (For example: `2` corresponds to 200%, `0.5` would
#'   be 50%.)
#' @param as Renderer.
#' @param ... Additional parameters passed to `ugd(...)`
#'
#' @return Rendered plot. Text renderers return strings, binary renderers
#'   return byte arrays.
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd_render_inline({
#'   hist(rnorm(100))
#' }, as = "svgz")
#'
#' s <- ugd_render_inline({
#'   plot.new()
#'   lines(c(0.5, 1, 0.5), c(0.5, 1, 1))
#' })
#' cat(s)
#' }
ugd_render_inline <- function(code,
                       page = 0,
                       width = getOption("unigd.width", 720),
                       height = getOption("unigd.height", 576),
                       zoom = 1,
                       as = "svg",
                       ...) {
  ugd(
    width = (width / zoom),
    height = (height / zoom),
    ...
  )
  tryCatch(code,
    finally = {
      tryCatch({
        s <- ugd_render(
          page = page,
          width = width,
          height = height,
          zoom = zoom,
          as = as
        )
      }, finally = {
        dev.off()
      })
    }
  )
  s
}


#' Inline plot rendering to a file.
#'
#' Convenience function for quick inline plot rendering.
#' This is similar to [ugd_save()] but the plotting code
#' is specified inline and an unigd graphics device is managed
#' (created and closed) automatically. Starting a device with [ugd()] is
#' therefore not necessary.
#'
#' @param code Plotting code. See examples for more information.
#' @param file Filepath to save plot.
#' @param page Plot page to render. If this is set to `0`, the last page will
#'   be selected. Can be set to a numeric plot index or plot ID
#'   (see [ugd_id()]).
#' @param width Width of the plot.
#' @param height Height of the plot.
#' @param zoom Zoom level. (For example: `2` corresponds to 200%, `0.5` would
#'   be 50%.)
#' @param as Renderer.
#' @param ... Additional parameters passed to `ugd(...)`
#'
#' @export
#'
#' @examples
#' \dontrun{
#'
#' ugd_save_inline({
#'   plot.new()
#'   lines(c(0.5, 1, 0.5), c(0.5, 1, 1))
#' }, file = "plot.svg")
#' }
ugd_save_inline <- function(code,
                       file,
                       page = 0,
                       width = getOption("unigd.width", 720),
                       height = getOption("unigd.height", 576),
                       zoom = 1,
                       as = "auto",
                       ...) {
  ugd(
    width = (width / zoom),
    height = (height / zoom),
    ...
  )
  tryCatch(code,
    finally = {
      tryCatch({
        ugd_save(
          file = file,
          page = page,
          width = width,
          height = height,
          zoom = zoom,
          as = as
        )
      }, finally = {
        dev.off()
      })
    }
  )
}

File: src/unigd.cpp
#include <cpp11/function.hpp>
#include <cpp11/strings.hpp>
#include <cpp11/list.hpp>
#include <cpp11/integers.hpp>
#include <cpp11/logicals.hpp>
#include <cpp11/data_frame.hpp>
#include <cpp11/as.hpp>
#include <cpp11/raws.hpp>

#include <vector>
#include <string>
#include <memory>

#include "unigd_dev.h"
#include "uuid.h"
#include "unigd_version.h"
#include "renderer_svg.h"
#include "renderers.h"
#include "r_thread.h"
#include "generic_dev.h"
#include "debug_print.h"

namespace
{
    inline std::shared_ptr<unigd::unigd_device> validate_unigddev(int devnum)
    {
        auto a = unigd::unigd_device::from_device_number(devnum);
        if (a == nullptr)
        {
            cpp11::stop("Not a valid device number");
        }
        return a;
    }

    inline long validate_plotid(const std::string &id)
    {
        try {
            return std::stol(id);
        }
        catch (const std::exception &e){
            cpp11::stop("Not a valid plot ID.");
        }
        return -1;
    }

} // namespace unigd


[[cpp11::register]]
int unigd_ugd_(std::string bg, double width, double height,
             double pointsize, cpp11::list aliases,
             bool reset_par)
{
    int ibg = R_GE_str2col(bg.c_str());

    const unigd::device_params dparams {
        ibg,
        width,
        height,
        pointsize,
        aliases,
        reset_par
    };

    return std::make_shared<unigd::unigd_device>(dparams)->create("unigd");
}

[[cpp11::register]]
cpp11::list unigd_state_(int devnum)
{
    auto dev = validate_unigddev(devnum);

    const auto state = dev->plt_state();

    SEXP client_status;
    /*std::shared_ptr<unigd::graphics_client> client;
    if (dev->get_client(&client))
    {
        client_status = cpp11::writable::strings({ client->client_status() });
    }
    else
    {*/
        client_status = R_NilValue;
    //}

    using namespace cpp11::literals;
    return cpp11::writable::list{
        "hsize"_nm = state.hsize,
        "upid"_nm = state.upid,
        "active"_nm = state.active,
        "client"_nm = client_status};
}

[[cpp11::register]]
cpp11::list unigd_info_(int devnum)
{
    /*auto dev = validate_unigddev(devnum);*/

    using namespace cpp11::literals;
    return cpp11::writable::list{
        "version"_nm = cpp11::writable::list{
        "unigd"_nm = UNIGD_VERSION,
        "cairo"_nm = UNIGD_VERSION_CAIRO
        }
    };
}

[[cpp11::register]]
cpp11::data_frame unigd_renderers_()
{
    using namespace cpp11::literals;

    const auto renderers = unigd::renderers::renderers();

    cpp11::writable::list rens{static_cast<R_xlen_t>(renderers->size())};

    const R_xlen_t nren = renderers->size();
    cpp11::writable::strings ren_id{nren};
    cpp11::writable::strings ren_mime{nren};
    cpp11::writable::strings ren_ext{nren};
    cpp11::writable::strings ren_name{nren};
    cpp11::writable::strings ren_type{nren};
    cpp11::writable::logicals ren_text;
    ren_text.resize(nren); // R cpp11 bug?
    cpp11::writable::strings ren_descr{nren};

    R_xlen_t i = 0;
    for (auto it = renderers->begin(); it != renderers->end(); it++)
    {
        ren_id[i] = it->second.info.id;
        ren_mime[i] = it->second.info.mime;
        ren_ext[i] = it->second.info.fileext;
        ren_name[i] = it->second.info.name;
        ren_type[i] = it->second.info.type;
        ren_text[i] = it->second.info.text;
        ren_descr[i] = it->second.info.description;
        i++;
    }

    cpp11::writable::data_frame res({
                "id"_nm = ren_id,
                "mime"_nm = ren_mime,
                "ext"_nm = ren_ext,
                "name"_nm = ren_name,
                "type"_nm = ren_type,
                "text"_nm = ren_text,
                "descr"_nm = ren_descr
    });
    return res;
}


[[cpp11::register]]
int unigd_plot_find_(int devnum, std::string plot_id)
{
    long pid = validate_plotid(plot_id);
    auto dev = validate_unigddev(devnum);
    auto page = dev->plt_index(pid);
    if (page == -1)
    {
        cpp11::stop("Not a valid plot ID.");
    }
    return page;
}

[[cpp11::register]]
SEXP unigd_render_(int devnum, int page, double width, double height, double zoom, std::string renderer_id)
{
    auto dev = validate_unigddev(devnum);

    if (width < 0 || height < 0)
    {
        zoom = 1;
    }

    unigd::renderers::renderer_map_entry ren;
    auto fi_renderer = unigd::renderers::find(renderer_id, &ren);
    if (!fi_renderer)
    {
        cpp11::stop("Not a valid string renderer ID.");
    }
    auto renderer = ren.generator();
    dev->plt_render(page, width / zoom, height / zoom, renderer.get(), zoom);

    const uint8_t *buf;
    size_t buf_size;
    renderer->get_data(&buf, &buf_size);

    if (ren.info.text) {
        return cpp11::writable::strings({ cpp11::r_string(std::string(buf, buf+buf_size)) });
    } else {
        return cpp11::writable::raws(buf, buf+buf_size);
    }
}

[[cpp11::register]]
bool unigd_remove_(int devnum, int page)
{
    auto dev = validate_unigddev(devnum);
    return dev->plt_remove(page);
}

[[cpp11::register]]
bool unigd_remove_id_(int devnum, std::string id)
{
    long pid = validate_plotid(id);
    auto dev = validate_unigddev(devnum);
    auto page = dev->plt_index(pid);
    if (page == -1)
    {
        cpp11::stop("Not a valid plot ID.");
    }

    return dev->plt_remove(page);
}

[[cpp11::register]]
cpp11::writable::list unigd_id_(int devnum, int page, int limit)
{
    auto dev = validate_unigddev(devnum);
    unigd::ex::find_results res;

    if (page == -1)
    {
        res = dev->plt_query_index(page);
    }
    else
    {
        res = dev->plt_query_range(page, limit);
    }

    using namespace cpp11::literals;
    cpp11::writable::list state{
        "hsize"_nm = res.state.hsize,
        "upid"_nm = res.state.upid,
        "active"_nm = res.state.active};

    cpp11::writable::list plots{static_cast<R_xlen_t>(res.ids.size())};

    for (std::size_t i = 0; i < res.ids.size(); ++i)
    {
        cpp11::writable::list p{"id"_nm = res.ids[i]};
        p.attr("class") = "unigd_pid";
        plots[i] = p;
    }

    return {
        "state"_nm = state,
        "plots"_nm = plots
    };
}

[[cpp11::register]]
bool unigd_clear_(int devnum)
{
    auto dev = validate_unigddev(devnum);
    return dev->plt_clear();
}

[[cpp11::register]]
void unigd_ipc_open_()
{
    unigd::async::ipc_open();
}

[[cpp11::register]]
void unigd_ipc_close_()
{
    unigd::async::ipc_close();
}
